### ISSUE TITLE
Particle system issues with Canvas+ and elsewhere

### ISSUE THREAD

**Issue by Leander1P** [playcanvas/engine#982]
We're developing an app for use in the Cocoon Developer App, Canvas+ mode, and we've noticed that a number of older phones and tablets seem to have issues with particles not showing up at all.  

There are also similar/identical issues outside of Cocoon's Canvas+ (in system webview and in some cases Safari/Chrome without any Cocoon involvement), but we don't have a great list of where yet -- I'll dump what we do know near the end of this issue.

I took an iPad 2 on iOS 8.4.1 and another iPad 2 on iOS 9.3.3, running the latest App Store version of Cocoon Developer App (2.4.0) and bisected, using `playcanvas-engine/examples/particle_system/sparks`.  It appears v0.182.0 is good (albeit a bit slow) on these devices and v0.183.0 is bad (no display).  Further investigation with git bisect:

```
good v0.182.0
bad v0.183.0 [note: FRAMEBUFFER_UNSUPPORTED]
bad [846acaf6330592735d03d82a825da182a95c61f1] fix culling again
good [f64c4356d00135704c006a17969ed59005188fc8] cleaning up [note: particles come out in batches and ~2FPS]
bad [707b6450b0e2f218e3f3318295b4fe362691cb1c] simplify output masks
bad [95f70e5002ae4da6c0d5317f92a443fb0349f728] cleaning up 4 [note: shader compile fail]
good [f62462bb83d53b1286271f98cec9f07cbcea44f8] fix position glitching [note: particles come out in batches and ~2FPS]
bad [3234f54e71d9b6691982a7034d983bffe64802ab] cleaning up 3 [note: shader compile fail]
bad [989bca83a73e0af0272ff980f41f280182c7c330] cleaning up 3 [note: shader compile fail] [git says: "first bad commit"]
```

So it started somewhere in the realm of 989bca83a73e0af0272ff980f41f280182c7c330 , although the shaders seemed to be in an unstable state for a bit around there (probably should have done `git bisect skip` on those).

Where I can figure out how to build with DEBUG enabled (after build.py changed to build.js =), I usually see a few instances of FRAMEBUFFER_UNSUPPORTED during startup for the "bad" commits, not sure if that is related.  `forceCpuParticles` doesn't *seem* to help (cribbing off https://forum.playcanvas.com/t/particles-issue-for-ipad-2-and-4/2237/3 ), although we've yet to be as rigorous with testing this.

Tagging @guycalledfrank since it seems to be related to a commit block where they were active; if you have any suggestions about things to try next or more info to look for, let us know (and thanks for the extra eyes in advance!).

Will keep peering at the changes near here and attempting to understand the nature of the changes and the problem, but thought it was worth getting an issue up in parallel.

----

We're *also* seeing some issues with particles wrapping when they shouldn't, similar to the above link, and some strangeness regarding particle scale and emitter position inconsistencies between devices -- but it seems like those should probably be saved for separate issues, as this blocks careful investigations of those (but if they're possibly related, can probably dig up more info here too).  Those issues seemed to be more dependent on forceCpuParticles, to some extent.

----

Other notes from earlier experiments, although note these were on custom projects and not as well-tracked -- they're here for posterity, but I wouldn't necessarily trust them without verification:

* "Repro: create a default scene, and add a default particle system. The particles are visible in PlayCanvas editor preview, on Android devices, and some iOS devices, but not these (in Cocoon, using either Canvas+ or Webview): iPhone 5, iOS 9.3.5 (bad), iPhone 4S, iOS 9.3.5 -- Note: enabling "this.app.graphicsDevice.forceCpuParticles" did not appear to resolve this issue."
* Test project was https://playcanv.as/p/VXn6EHTJ/
* "My tests on my 6S+ iOS 10.3.3 with Safari and Chrome both looked right, excepting that the particles are wrapping when maybe they shouldn't."
* "Confirmed it also does not work on my iPhone 5 + iOS 10.3.3 in Safari or whatever version of Chrome I have on here."
* (All of the above were before turning on "forceCpuParticles".  Test project URL above was changed at this point to turn on "forceCpuParticles" and "also removed the plane and cube, and
added a few more particle systems with a variety of sizes")
* "...and this doesn't seem to work on my iPod Touch 6... no, wait, only [works] in Webview, not [working in] Canvas+..."
* "...though still doesn't work in Webview or Canvas+ on iPhone 4S..."
* "...and I'm getting something very strange on Nexus 5 [in Canvas+, WebView looks fine]; I'm very confused..." (this was some kind of corruption that was causing many instances of the screen to repeat, with no particles evident, just the little playcanvas eye icon)

**Comment by willeastcott** [comment]
Hey @guycalledfrank, can you run TANX on your iPhone and check if the particles are displaying when you play? e.g. Shells exploding against walls. I don't see them on my iPhone 7. Perhaps this is related?

**Comment by Leander1P** [comment]
As a tiny status update, we may have a workaround for our particular issues (although it doesn't yet explain why things broke around then for sparks on the aforementioned devices).

Our attempt to set `forceCpuParticles` was botched somehow; if we force it to `true` in playcanvas-latest.js near the `unmaskedRenderer` branches in a near-HEAD build (rather than trying in a game script), the sparks example works, as does our demo (although positions or scales are messed up, will investigate this shortly) -- on at least an iPad 2 with 8.4.1 and an iPhone 5 with 10.3.3 (the only two I've tested on).  

So a workaround for our particular issues may come down to more `unmaskedRenderer` special-cases a la those suggested at https://forum.playcanvas.com/t/particles-issue-for-ipad-2-and-4/2237/3 .

I'm going to keep hunting for a bit to try to see if I can ascertain why we're getting some unexpected position/scale in our particle systems with `forceCpuParticles`, but that should probably end up in another issue or PR to avoid making this one any more vague than it already is.

Also may continue to try to figure out why things broke around the commits mentioned in the bisect above -- modulus temporary shader compile errors -- with the default forceCpuParticles (undefined/false, I think, on these devices?).  It would be nice to have GPU particles on these devices, if at all possible.  Will try to report back here with any findings, if time and deadlines allow any more investigation.

**Comment by willeastcott** [comment]
Looks like [Cocoon has now shut down](https://blog.cocoon.io/) so I'm going to close this issue. 😢
